### ISSUE TITLE
Can't use queryset.update to update a streamfield

### ISSUE THREAD

**Issue by seb-b** [wagtail/wagtail#2626]
Usually I setup some test data for each project which creates some pages and content.

The pattern I've been using is something like this:

```
page = PageModel(
   title='whatever'
   body='<p>hello</p>'
   sidebar=json.dumps([** code to create streamfield content **])
)
site_root.add_child(instance=page)
```

I had a model where I needed to comeback and update a streamfield after the page had been created.
Using this:

```
PageModel.objects.all().update(sidebar=json.dumps([** streamfield content **])
```

gives me the error 

```
Traceback (most recent call last):
  File "./manage.py", line 15, in <module>
    execute_from_command_line(args)
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/vagrant/dma/management/commands/testdata.py", line 166, in handle
    self.add_sidebar_content()
  File "/vagrant/dma/management/commands/testdata.py", line 321, in add_sidebar_content
    AdviceIndex.objects.all().update(sidebar_content=self.random_sidebar_blocks(3))
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/db/models/query.py", line 628, in update
    rows = query.get_compiler(self.db).execute_sql(CURSOR)
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/db/models/sql/compiler.py", line 1149, in execute_sql
    cursor = super(SQLUpdateCompiler, self).execute_sql(result_type)
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/db/models/sql/compiler.py", line 837, in execute_sql
    sql, params = self.as_sql()
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/db/models/sql/compiler.py", line 1117, in as_sql
    val = field.get_db_prep_save(val, connection=self.connection)
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/db/models/fields/__init__.py", line 728, in get_db_prep_save
    prepared=False)
  File "/home/vagrant/venv/lib/python3.4/site-packages/django/db/models/fields/__init__.py", line 720, in get_db_prep_value
    value = self.get_prep_value(value)
  File "/home/vagrant/venv/lib/python3.4/site-packages/wagtail/wagtailcore/fields.py", line 110, in get_prep_value
    return json.dumps(self.stream_block.get_prep_value(value), cls=DjangoJSONEncoder)
  File "/home/vagrant/venv/lib/python3.4/site-packages/wagtail/wagtailcore/blocks/stream_block.py", line 209, in get_prep_value
    for child in value  # child is a BoundBlock instance
  File "/home/vagrant/venv/lib/python3.4/site-packages/wagtail/wagtailcore/blocks/stream_block.py", line 209, in <listcomp>
    for child in value  # child is a BoundBlock instance
AttributeError: 'str' object has no attribute 'block'
```

If I rewrite the update to do the following it works fine though

```
for page in PageModel.objects.all()
    page.sidebar = json.dumps([** streamfield content **]
    page.save()
```

**Comment by gasman** [comment]
I *think* the 'error' here is that Wagtail is being more liberal than it ought to be when it allows you to assign a JSON string to `page.sidebar` in your last code snippet. If StreamField is doing its job properly, you should never have to work with JSON strings on the Python side: the JSON is only the database representation, and when you read (or write) the value through the model field, StreamField should translate it from JSON to a native Python value (and vice versa).

However, I suspect we're not currently being particularly rigorous about what that native Python value actually is. I think the intended behaviour is:

* reading back a value from a StreamField should give you a `StreamValue` instance (a list-like object that knows enough about the StreamField's configuration to make the HTML rendering work)
* when writing back to a StreamField, you can pass it a `StreamValue`, or a list of (block_type_identifier, value) tuples. (The latter option is needed because `StreamValue` objects need a reference back to the StreamBlock, which is fiddly to set up and sometimes impossible, e.g. when setting a `default` value on the StreamField.) As per Postel's Law, accepting any other representations (such as a JSON string) would be a bonus, but only if we can make it work consistently and non-confusingly...

I'm not sure how close we currently are to that behaviour - working that out will probably involve untangling the `get_prep_value` / `to_python` logic, which we'll tackle in #2119. In the meantime, I'd suggest that you try passing a list of tuples, rather than a JSON string:

    PageModel.objects.all().update(sidebar=[
        ('title', 'Hello world'),
        ('paragraph', '<p>Lorem ipsum dolor sit amet...</p>')
    ])

I can't guarantee that will work, but I'm fairly sure it's the version of the code that we *want* to make work :-) )

**Comment by thibaudcolas** [comment]
Reviewed today – assigning to @gasman to check if this might have been addressed already (Wagtail 3.0?)
